### PR TITLE
Secret result

### PR DESCRIPTION
**Pull request by cicdw**
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
Adds a new SecretResult type for handling sensitive data that tasks might return - depends on #2302 

## Why is this PR important?
PIN 16 work

### CODE REVIEW

File: src/prefect/engine/results/secret_result.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            - Result: a new result instance with the data represented by the filepath
+        """
+        new = self.copy()
+        new.value = self.secret_task.run(name=filepath)
+        new.filepath = filepath
+        return new
+
+    def write(self, value: Any, **kwargs: Any) -> Result:
+        """
+        .value` and returns `self`.
```
Comment (joshmeek) [inline src/prefect/engine/results/secret_result.py:41]: Missing docstring text

File: src/prefect/engine/results/secret_result.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            - **kwargs (optional): unused, for interface compatibility
+
+        Raises:
+            - ValueError: SecretResults cannot be written to
+        """
+        raise ValueError("SecretResults cannot be written to.")
+
+    def exists(self, filepath: str) -> bool:
+        """
+        Confirms that the provided value is JSON deserializable.
```
Comment (joshmeek) [inline src/prefect/engine/results/secret_result.py:54]: PrefectResult docstring

Comment (cicdw) [reply src/prefect/engine/results/secret_result.py]: @joshmeek updated

File: src/prefect/engine/results/secret_result.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class SecretResult(Result):
+    """
+    Hook for storing and retrieving sensitive task results from a Secret store. Only
+    intended to be used for Secret Tasks - each call to "read" will actually rerun
+    the underlying secret task and re-retrieve the secret value.
+
+    Args:
+        - value (Any): the underlying value this Result should represent
```
Comment (joshmeek) [inline src/prefect/engine/results/secret_result.py:14]: No longer a kwarg

File: src/prefect/engine/results/secret_result.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    """
+
+    def __init__(
+        self, secret_task: "prefect.tasks.secrets.Secret", **kwargs: Any
+    ) -> None:
+        self.secret_task = secret_task
+        kwargs.setdefault("filepath", secret_task.name)
+        super().__init__(**kwargs)
+
+    def read(self, filepath: str) -> Result:
```
Comment (joshmeek) [inline src/prefect/engine/results/secret_result.py:24]: Update it to location while you're here :)

Comment (cicdw) [reply src/prefect/engine/results/secret_result.py:24]: Nooooooo

**Pull request merged**